File: backend/src/backend/schemas/identification.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class IdentificationJobCreate(BaseModel):
    song_ids: list[int] = Field(default_factory=list, description="V1 song ids to identify")
    file_ids: list[int] = Field(default_factory=list, description="V2 file ids to identify")


class CandidateRead(BaseModel):
    id: int
    file_id: int
    source: str
    score: Decimal
    confidence_level: str
    status: str
    recording_mbid: str | None = None
    title: str | None = None
    artist: str | None = None
    release_title: str | None = None

    model_config = ConfigDict(from_attributes=False)


class IdentificationJobRead(BaseModel):
    id: int
    job_type: str
    status: str
    progress: float
    result_json: dict | list | None = None
    error_message: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ReleaseMatchRead(BaseModel):
    release_id: int
    release_mbid: str | None = None
    release_title: str | None = None
    track_number: int | None = None
    disc_number: int | None = None


class AcceptResponse(BaseModel):
    accepted: bool
    file_id: int
    resolved: dict[str, str | None] | None = None
    recording_id: int | None = None
    release: ReleaseMatchRead | None = None


class ReviewCandidateRead(BaseModel):
    """Pending candidate joined with its file for the review queue (§22)."""

    id: int
    file_id: int
    filename: str
    filepath: str
    source: str
    score: Decimal
    confidence_level: str
    status: str
    recording_mbid: str | None = None
    title: str | None = None
    artist: str | None = None
    release_title: str | None = None


class BulkAcceptRequest(BaseModel):
    """Either explicit candidate ids, or a filter accepting the best pending
candidate per file (§22 bulk review)."""

    candidate_ids: list[int] = Field(default_factory=list)
    confidence_level: str | None = None
    min_score: float | None = None


class BulkAcceptResponse(BaseModel):
    accepted: list[AcceptResponse]
    skipped: int = 0
    errors: list[str] = Field(default_factory=list)


class ReleasePreferences(BaseModel):
    preference: str
    country: str = ""
    label: str = ""

File: frontend/src/lib/api/identification.ts
import { api } from './client';

export interface IdentificationJob {
	id: number;
	job_type: 'identify';
	status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
	progress: number;
	result_json?: Record<string, unknown> | null;
	error_message?: string | null;
	created_at: string;
	completed_at?: string | null;
}

export interface IdentificationCandidate {
	id: number;
	file_id: number;
	source: string;
	score: number;
	confidence_level: 'auto_apply' | 'auto_suggest_review' | 'review_required' | 'no_match';
	status: 'pending' | 'accepted' | 'rejected' | 'expired';
	recording_mbid?: string | null;
	title?: string | null;
	artist?: string | null;
	release_title?: string | null;
}

export async function startIdentificationJob(songIds: number[]): Promise<IdentificationJob> {
	return api.post<IdentificationJob>('/identification/jobs', { song_ids: songIds });
}

export async function getIdentificationJob(jobId: number): Promise<IdentificationJob> {
	return api.get<IdentificationJob>(`/identification/jobs/${jobId}`);
}

export async function identifySong(songId: number): Promise<IdentificationJob> {
	return api.post<IdentificationJob>(`/identification/songs/${songId}/identify`);
}

export async function getSongCandidates(songId: number): Promise<IdentificationCandidate[]> {
	return api.get<IdentificationCandidate[]>(`/identification/songs/${songId}/candidates`);
}

export async function acceptCandidate(
	songId: number,
	candidateId: number
): Promise<{ accepted: boolean; file_id: number; recording_id?: number | null }> {
	return api.post(`/identification/songs/${songId}/candidates/${candidateId}/accept`);
}

export async function rejectCandidate(songId: number, candidateId: number): Promise<void> {
	return api.post(`/identification/songs/${songId}/candidates/${candidateId}/reject`);
}

export interface ReviewCandidate {
	id: number;
	file_id: number;
	filename: string;
	filepath: string;
	source: string;
	score: number;
	confidence_level: 'auto_apply' | 'auto_suggest_review' | 'review_required' | 'no_match';
	status: 'pending' | 'accepted' | 'rejected' | 'expired';
	recording_mbid?: string | null;
	title?: string | null;
	artist?: string | null;
	release_title?: string | null;
}

export interface BulkAcceptResult {
	accepted: unknown[];
	skipped: number;
	errors: string[];
}

export async function getReviewQueue(params?: {
	confidence_level?: string;
	source?: string;
	min_score?: number;
}): Promise<ReviewCandidate[]> {
	const query = new URLSearchParams();
	if (params?.confidence_level) query.set('confidence_level', params.confidence_level);
	if (params?.source) query.set('source', params.source);
	if (params?.min_score != null) query.set('min_score', String(params.min_score));
	const qs = query.toString();
	return api.get(`/identification/review${qs ? `?${qs}` : ''}`);
}

export async function bulkAccept(body: {
	candidate_ids?: number[];
	confidence_level?: string;
	min_score?: number;
}): Promise<BulkAcceptResult> {
	return api.post('/identification/review/bulk-accept', body);
}

export interface ReleasePreferences {
	preference: string;
	country: string;
	label: string;
}

export async function getReleasePreferences(): Promise<ReleasePreferences> {
	return api.get('/identification/release-preferences');
}

export async function updateReleasePreferences(
	prefs: ReleasePreferences
): Promise<ReleasePreferences> {
	return api.put('/identification/release-preferences', prefs);
}
